drivers/gpu/drm/v3d/v3d_drv.c

```diff
@@ -232,17 +267,40 @@ static int v3d_platform_drm_probe(struct platform_device *pdev)
 	if (ret)
 		return ret;
 
+	if (v3d->ver >= V3D_GEN_71) {
+		ret = map_regs(v3d, &v3d->sms_regs, "sms");
+		if (ret)
+			return ret;
+	}
+
+	v3d->clk = devm_clk_get_optional(dev, NULL);
+	if (IS_ERR(v3d->clk))
+		return dev_err_probe(dev, PTR_ERR(v3d->clk), "Failed to get V3D clock\n");
+
+	ret = clk_prepare_enable(v3d->clk);
+	if (ret) {
+		dev_err(&pdev->dev, "Couldn't enable the V3D clock\n");
+		return ret;
+	}
+
+	v3d_idle_sms(v3d);
+
 	mmu_debug = V3D_READ(V3D_MMU_DEBUG_INFO);
 	mask = DMA_BIT_MASK(30 + V3D_GET_FIELD(mmu_debug, V3D_MMU_PA_WIDTH));
 	ret = dma_set_mask_and_coherent(dev, mask);
 	if (ret)
-		return ret;
+		goto clk_disable;
 
 	v3d->va_width = 30 + V3D_GET_FIELD(mmu_debug, V3D_MMU_VA_WIDTH);
 
 	ident1 = V3D_READ(V3D_HUB_IDENT1);
 	v3d->ver = (V3D_GET_FIELD(ident1, V3D_HUB_IDENT1_TVER) * 10 +
```

Comment:
It looks strange to change this critical value, even with the WARN. Would you consider using `gen` before here, and `v3d->ver` after?

Reply:
How do you feel about this approach?

```c
v3d->ver = (enum v3d_gen)of_device_get_match_data(dev);

[...]

gen = (V3D_GET_FIELD(ident1, V3D_HUB_IDENT1_TVER) * 10 + V3D_GET_FIELD(ident1, V3D_HUB_IDENT1_REV));
WARN_ON(v3d->ver != gen);
```

This way we wouldn't change the value, but still check that the hardware returns the same value.

Reply:
I think the hardware is definitive, and the DT setting is an early hint, which is why I suggested doing it the other way round:
```c
	gen = (enum v3d_gen)of_device_get_match_data(dev);

[...]

	if (gen >= V3D_GEN_71) {
		ret = map_regs(v3d, &v3d->sms_regs, "sms");
		if (ret)
			return ret;
	}

[...]

	v3d->ver = (V3D_GET_FIELD(ident1, V3D_HUB_IDENT1_TVER) * 10 + V3D_GET_FIELD(ident1, V3D_HUB_IDENT1_REV));
	WARN_ON(v3d->ver != gen);
```

Reply:
The thing is: between `map_regs` and `V3D_GET_FIELD`, we have `v3d_sms_idle()`, which uses `v3d->ver`

 I'm sorry, but I don't understand what the issue is. First, we assign `v3d->ver = gen` and after we power up the GPU, we overwrite the variable with the hardware value.

In the end, `v3d->ver` will be the one from the hardware, even if `v3d->ver != gen`.